src/dbnode/storage/series/reader.go

```diff
@@ -119,88 +119,176 @@ func (r *Reader) readersWithBlocksMapAndBuffer(
 		blockSize   = ropts.BlockSize()
 		readerCount = end.Sub(start) / blockSize
+		buffer      [][]xio.BlockReader
+		cached      []xio.BlockReader
 	)
 
 	if readerCount < 0 {
 		readerCount = 0
 	}
 
-	// Two-dimensional slice such that the first dimension is unique by blockstart
-	// and the second dimension is blocks of data for that blockstart (not necessarily
-	// in chronological order).
-	//
-	// ex. (querying 2P.M -> 6P.M with a 2-hour blocksize):
-	// [][]xio.BlockReader{
-	//   {block0, block1, block2}, // <- 2P.M
-	//   {block0, block1}, // <-4P.M
-	// }
-	results := make([][]xio.BlockReader, 0, readerCount)
 	for blockAt := start; !blockAt.After(end); blockAt = blockAt.Add(blockSize) {
-		// resultsBlock holds the results from one block. The flow is:
-		// 1) Look in the cache for metrics for a block.
-		// 2) If there is nothing in the cache, try getting metrics from disk.
-		// 3) Regardless of (1) or (2), look for metrics in the series buffer.
+		// Eagerly load the readers from the series buffer and disk cache to avoid holding locks in the iterator.
 		//
 		// It is important to look for data in the series buffer one block at
 		// a time within this loop so that the returned results contain data
 		// from blocks in chronological order. Failure to do this will result
 		// in an out of order error in the MultiReaderIterator on query.
-		var resultsBlock []xio.BlockReader
 
-		blockReader, block, found, err := retrieveCached(ctx, blockAt, seriesBlocks)
+		blockReader, blk, found, err := retrieveCached(ctx, blockAt, seriesBlocks)
 		if err != nil {
 			return nil, err
 		}
 
 		if found {
 			// NB(r): Mark this block as read now
-			block.SetLastReadTime(now)
+			blk.SetLastReadTime(now)
 			if r.onRead != nil {
-				r.onRead.OnReadBlock(block)
+				r.onRead.OnReadBlock(blk)
 			}
-		} else {
-			blockReader, found, err = r.streamBlock(ctx, blockAt, r.onRetrieve, nsCtx)
-			if err != nil {
-				return nil, err
-			}
-		}
-
-		if found {
-			resultsBlock = append(resultsBlock, blockReader)
+			cached = append(cached, blockReader)
 		}
 
 		if seriesBuffer != nil {
+			var bufferReaders []xio.BlockReader
 			bufferResults, err := seriesBuffer.ReadEncoded(ctx, blockAt, blockAt.Add(blockSize), nsCtx)
 			if err != nil {
 				return nil, err
 			}
+
 			// Multiple block results may be returned here (for the same block
 			// start) - one for warm writes and another for cold writes.
 			for _, bufferRes := range bufferResults {
-				resultsBlock = append(resultsBlock, bufferRes...)
+				bufferReaders = append(bufferReaders, bufferRes...)
+			}
+			if len(bufferReaders) > 0 {
+				buffer = append(buffer, bufferReaders)
 			}
-		}
-
-		if len(resultsBlock) > 0 {
-			results = append(results, resultsBlock)
 		}
 	}
 
-	return results, nil
+	return &blockReaderIter{
+		start:     start,
+		end:       end,
+		blockSize: blockSize,
+		reader:    r,
+		nsCtx:     nsCtx,
+		buffer:    buffer,
+		cached:    cached,
+	}, nil
```

Comment:
nit: Would probably be nicer if we we used a `newBlockReaderIter(blockReaderIterOptions{....})` for option evolution in the future and separation of arguments vs state on the block reader iter, but minor thing.

Reply:
If this is the only thing that would require a followup change let's leave it for later though, this is all local to this file, a little less of an issue as such.